File: dotnet/Zirve.Sdk/Registry/RegistryManager.cs
using System;
using System.Net.Http;
using System.Text.Json;
using System.Threading.Tasks;
using Zirve.Sdk.Config;

namespace Zirve.Sdk.Registry;

public class RegistryManager
{
    private readonly HttpClient _httpClient;
    private readonly string _url;
    private readonly string _authHeader;

    public RegistryManager(HttpClient httpClient, ConfigManager configManager)
    {
        _httpClient = httpClient ?? throw new ArgumentNullException(nameof(httpClient));
        
        var cfg = configManager.Module("registry");
        _url = (cfg.GetValueOrDefault("url", "http://harbor-core.zirve-infra.svc.cluster.local") ?? "").TrimEnd('/') + "/api/v2.0";
        
        var user = cfg.GetValueOrDefault("username", "admin");
        var pass = cfg.GetValueOrDefault("password", "Harbor12345");
        
        var rawAuth = System.Text.Encoding.ASCII.GetBytes($"{user}:{pass}");
        _authHeader = "Basic " + Convert.ToBase64String(rawAuth);
    }

    private async Task<JsonElement> RequestAsync(HttpMethod method, string path, object? body = null)
    {
        var request = new HttpRequestMessage(method, $"{_url}/{path.TrimStart('/')}");
        request.Headers.Add("Authorization", _authHeader);
        
        if (body != null)
        {
            var json = body is string s ? s : JsonSerializer.Serialize(body);
            request.Content = new StringContent(json, System.Text.Encoding.UTF8, "application/json");
        }

        var response = await _httpClient.SendAsync(request);
        
        if (response.StatusCode == System.Net.HttpStatusCode.NotFound)
            return default;

        response.EnsureSuccessStatusCode();

        if (response.StatusCode == System.Net.HttpStatusCode.NoContent || response.Content.Headers.ContentLength == 0)
            return default;

        using var doc = await JsonDocument.ParseAsync(await response.Content.ReadAsStreamAsync());
        return doc.RootElement.Clone();
    }

    public async Task<JsonElement> ListProjectsAsync()
    {
        return await RequestAsync(HttpMethod.Get, "projects");
    }

    public async Task<JsonElement> ListImagesAsync(string projectName)
    {
        return await RequestAsync(HttpMethod.Get, $"projects/{Uri.EscapeDataString(projectName)}/repositories");
    }

    public async Task<bool> ScanImageAsync(string projectName, string repositoryName, string reference)
    {
        try
        {
            var safeRepo = Uri.EscapeDataString(repositoryName);
            await RequestAsync(HttpMethod.Post, $"projects/{Uri.EscapeDataString(projectName)}/repositories/{safeRepo}/artifacts/{Uri.EscapeDataString(reference)}/scan");
            return true;
        }
        catch
        {
            return false;
        }
    }

    public async Task<JsonElement> ScanReportAsync(string projectName, string repositoryName, string reference)
    {
        var safeRepo = Uri.EscapeDataString(repositoryName);
        return await RequestAsync(HttpMethod.Get, $"projects/{Uri.EscapeDataString(projectName)}/repositories/{safeRepo}/artifacts/{Uri.EscapeDataString(reference)}/additions/vulnerabilities");
    }

    public async Task<bool> HealthAsync()
    {
        try
        {
            var response = await _httpClient.GetAsync($"{_url}/health");
            return response.IsSuccessStatusCode;
        }
        catch
        {
            return false;
        }
    }
}

File: typescript/src/Registry/RegistryManager.ts
/**
 * Zirve Registry Manager — Harbor
 *
 * Integrates with Harbor v2 HTTP API to list images, scan vulnerabilities, and manage tags.
 */
export class RegistryManager {
  private config: Record<string, any>;
  private url: string;
  private authHeader: string;

  constructor(config: Record<string, any>) {
    this.config = config;
    this.url = (config.url || 'http://harbor-core.zirve-infra.svc.cluster.local').replace(/\/$/, '') + '/api/v2.0';
    
    const user = config.username || 'admin';
    const pass = config.password || 'Harbor12345';
    this.authHeader = 'Basic ' + Buffer.from(`${user}:${pass}`).toString('base64');
  }

  private async request(method: string, path: string, body?: any): Promise<any> {
    const options: RequestInit = {
      method,
      headers: {
        'Authorization': this.authHeader,
        'Content-Type': 'application/json'
      },
      body: body ? JSON.stringify(body) : undefined
    };

    const response = await fetch(`${this.url}/${path.replace(/^\/+/, '')}`, options);
    
    if (!response.ok) {
      if (response.status === 404) return null;
      throw new Error(`Harbor API Error [${response.status}]`);
    }

    if (response.status === 204 || response.headers.get('content-length') === '0') return true;
    return await response.json();
  }

  public async listProjects(): Promise<any[]> {
    const projects = await this.request('GET', 'projects');
    return projects || [];
  }

  public async listImages(projectName: string): Promise<any[]> {
    const repos = await this.request('GET', `projects/${projectName}/repositories`);
    return repos || [];
  }

  public async scanImage(projectName: string, repositoryName: string, reference: string): Promise<boolean> {
    // Escaping repo name required by Harbor for deeply nested paths
    const safeRepo = encodeURIComponent(repositoryName);
    try {
      await this.request('POST', `projects/${projectName}/repositories/${safeRepo}/artifacts/${reference}/scan`);
      return true;
    } catch {
      return false;
    }
  }

  public async scanReport(projectName: string, repositoryName: string, reference: string): Promise<any> {
    const safeRepo = encodeURIComponent(repositoryName);
    return await this.request('GET', `projects/${projectName}/repositories/${safeRepo}/artifacts/${reference}/additions/vulnerabilities`);
  }

  public async health(): Promise<boolean> {
    try {
      const response = await fetch(`${this.url}/health`);
      return response.ok;
    } catch {
      return false;
    }
  }
}
